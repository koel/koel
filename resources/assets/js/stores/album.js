import _ from 'lodash';

import utils from '../services/utils';
import stub from '../stubs/album';
import songStore from './song';

export default {
    stub,
    artists: [],
    
    state: {
        albums: [stub],    
    },

    /**
     * Init the store.
     * 
     * @param  array artists The array of artists to extract album data from.
     */
    init(artists) {
        this.artists = artists;

        // Traverse through the artists array and add their albums into our master album list.
        this.state.albums = _.reduce(artists, (albums, artist) => {
            // While we're doing so, for each album, we get its length
            // and keep a back reference to the artist too.
            _.each(artist.albums, album => {
                album.artist = artist;
                this.getLength(album);
            });

            return albums.concat(artist.albums);
        }, []);

        // Then we init the song store.
        songStore.init(this.state.albums);
    },

    all() {
        return this.state.albums;
    },

    /**
     * Get the total length of an album by summing up its songs' duration.
     * The length will also be converted into a H:i:s format and stored as fmtLength.
     * 
     * @param  object album
     * 
     * @return string The H:i:s format of the album.
     */
    getLength(album) {
        album.length = _.reduce(album.songs, (length, song) => length + song.length, 0);

        return (album.fmtLength = utils.secondsToHis(album.length));
    },
};
