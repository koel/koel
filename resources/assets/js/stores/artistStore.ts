import { reactive, UnwrapNestedRefs } from 'vue'
import { differenceBy, unionBy } from 'lodash'
import { cache, http } from '@/services'
import { arrayify, logger } from '@/utils'

const UNKNOWN_ARTIST_ID = 1
const VARIOUS_ARTISTS_ID = 2

export const artistStore = {
  vault: new Map<number, UnwrapNestedRefs<Artist>>(),

  state: reactive({
    artists: [] as Artist[]
  }),

  byId (id: number) {
    return this.vault.get(id)
  },

  removeByIds (ids: number[]) {
    this.state.artists = differenceBy(this.state.artists, ids.map(id => this.byId(id)), 'id')
    ids.forEach(id => this.vault.delete(id))
  },

  isVarious: (artist: Artist | number) => (typeof artist === 'number')
    ? artist === VARIOUS_ARTISTS_ID
    : artist.id === VARIOUS_ARTISTS_ID,

  isUnknown: (artist: Artist | number) => (typeof artist === 'number')
    ? artist === UNKNOWN_ARTIST_ID
    : artist.id === UNKNOWN_ARTIST_ID,

  isStandard (artist: Artist | number) {
    return !this.isVarious(artist) && !this.isUnknown(artist)
  },

  async uploadImage (artist: Artist, image: string) {
    artist.image = (await http.put<{ imageUrl: string }>(`artist/${artist.id}/image`, { image })).imageUrl

    // sync to vault
    this.byId(artist.id)!.image = artist.image

    return artist.image
  },

  syncWithVault (artists: Artist | Artist[]) {
    return arrayify(artists).map(artist => {
      let local = this.vault.get(artist.id)
      local = local ? Object.assign(local, artist) : reactive(artist)
      this.vault.set(artist.id, local)

      return local
    })
  },

  async resolve (id: number) {
    let artist = this.byId(id)

    if (!artist) {
      try {
        artist = this.syncWithVault(
          await cache.remember<Artist>(['artist', id], async () => await http.get<Artist>(`artists/${id}`))
        )[0]
      } catch (e) {
        logger.error(e)
      }
    }

    return artist
  },

  async paginate (page: number) {
    const resource = await http.get<PaginatorResource>(`artists?page=${page}`)
    this.state.artists = unionBy(this.state.artists, this.syncWithVault(resource.data), 'id')

    return resource.links.next ? ++resource.meta.current_page : null
  }
}
