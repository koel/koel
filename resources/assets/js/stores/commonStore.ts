import isMobile from 'ismobilejs'
import { reactive } from 'vue'
import { http } from '@/services'
import { playlistFolderStore, playlistStore, preferenceStore, queueStore, settingStore, themeStore, userStore } from '.'

interface CommonStoreState {
  allow_download: boolean
  cdn_url: string
  current_user: User
  current_version: string
  latest_version: string
  playlists: Playlist[]
  playlist_folders: PlaylistFolder[]
  settings: Settings
  use_i_tunes: boolean
  use_last_fm: boolean
  use_spotify: boolean
  users: User[]
  use_you_tube: boolean,
  song_count: number,
  song_length: number,
  queue_state: QueueState
}

export const commonStore = {
  state: reactive<CommonStoreState>({
    allow_download: false,
    cdn_url: '',
    current_user: undefined as unknown as User,
    current_version: '',
    latest_version: '',
    playlists: [],
    playlist_folders: [],
    settings: {} as Settings,
    use_i_tunes: false,
    use_last_fm: false,
    use_spotify: false,
    users: [],
    use_you_tube: false,
    song_count: 0,
    song_length: 0,
    queue_state: {
      type: 'queue-states',
      songs: [],
      current_song: null,
      playback_position: 0
    }
  }),

  async init () {
    Object.assign(this.state, await http.get<CommonStoreState>('data'))

    // Always disable YouTube integration on mobile.
    this.state.use_you_tube = this.state.use_you_tube && !isMobile.phone

    // If this is a new user, initialize his preferences to be an empty object.
    this.state.current_user.preferences = this.state.current_user.preferences || {}

    userStore.init(this.state.current_user)
    preferenceStore.init(this.state.current_user)
    playlistStore.init(this.state.playlists)
    playlistFolderStore.init(this.state.playlist_folders)
    settingStore.init(this.state.settings)
    queueStore.init(this.state.queue_state)
    themeStore.init()

    return this.state
  }
}
