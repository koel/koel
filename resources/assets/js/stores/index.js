export * from "./album";
export * from "./artist";
export * from "./equalizer";
export * from "./favorite";
export * from "./playlist";
export * from "./preference";
export * from "./queue";
export * from "./setting";
export * from "./shared";
export * from "./song";
export * from "./user";
