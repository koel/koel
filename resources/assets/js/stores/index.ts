export * from './albumStore'
export * from './artistStore'
export * from './commonStore'
export * from './equalizerStore'
export * from './favoriteStore'
export * from './overviewStore'
export * from './playlistStore'
export * from './playlistFolderStore'
export * from './preferenceStore'
export * from './queueStore'
export * from './recentlyPlayedStore'
export * from './searchStore'
export * from './settingStore'
export * from './songStore'
export * from './themeStore'
export * from './userStore'
export * from './genreStore'
export * from './visualizerStore'
