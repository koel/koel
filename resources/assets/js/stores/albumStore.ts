import { reactive, UnwrapNestedRefs } from 'vue'
import { differenceBy, merge, unionBy } from 'lodash'
import { cache, http } from '@/services'
import { arrayify, logger } from '@/utils'
import { songStore } from '@/stores'

const UNKNOWN_ALBUM_ID = 1

export const albumStore = {
  vault: new Map<number, UnwrapNestedRefs<Album>>(),

  state: reactive({
    albums: [] as Album[]
  }),

  byId (id: number) {
    return this.vault.get(id)
  },

  removeByIds (ids: number[]) {
    this.state.albums = differenceBy(this.state.albums, ids.map(id => this.byId(id)), 'id')
    ids.forEach(id => {
      this.vault.delete(id)
      cache.remove(['album', id])
    })
  },

  isUnknown: (album: Album | number) => {
    if (typeof album === 'number') return album === UNKNOWN_ALBUM_ID
    return album.id === UNKNOWN_ALBUM_ID
  },

  syncWithVault (albums: Album | Album[]) {
    return arrayify(albums).map(album => {
      let local = this.vault.get(album.id)
      local = reactive(local ? merge(local, album) : album)
      this.vault.set(album.id, local)

      return local
    })
  },

  /**
   * Upload a cover for an album.
   *
   * @param {Album} album The album object
   * @param {string} cover The content data string of the cover
   */
  async uploadCover (album: Album, cover: string) {
    album.cover = (await http.put<{ coverUrl: string }>(`album/${album.id}/cover`, { cover })).coverUrl
    songStore.byAlbum(album).forEach(song => song.album_cover = album.cover)

    // sync to vault
    this.byId(album.id)!.cover = album.cover

    return album.cover
  },

  /**
   * Fetch the (blurry) thumbnail-sized version of an album's cover.
   */
  fetchThumbnail: async (id: number) => {
    return (await http.get<{ thumbnailUrl: string }>(`album/${id}/thumbnail`)).thumbnailUrl
  },

  async resolve (id: number) {
    let album = this.byId(id)

    if (!album) {
      try {
        album = this.syncWithVault(
          await cache.remember<Album>(['album', id], async () => await http.get<Album>(`albums/${id}`))
        )[0]
      } catch (e) {
        logger.error(e)
      }
    }

    return album
  },

  async paginate (page: number) {
    const resource = await http.get<PaginatorResource>(`albums?page=${page}`)
    this.state.albums = unionBy(this.state.albums, this.syncWithVault(resource.data), 'id')

    return resource.links.next ? ++resource.meta.current_page : null
  },

  async fetchForArtist (artist: Artist | number) {
    const id = typeof artist === 'number' ? artist : artist.id

    return this.syncWithVault(
      await cache.remember<Album[]>(['artist-albums', id], async () => await http.get<Album[]>(`artists/${id}/albums`))
    )
  }
}
