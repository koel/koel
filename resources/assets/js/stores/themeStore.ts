import { clone, uniqBy } from 'lodash'
import Color from 'color'
import StyleObserver from 'style-observer'
import { reactive } from 'vue'
import { preferenceStore as preferences } from '@/stores/preferenceStore'
import { http } from '@/services/http'
import themes from '@/themes'
import { cache } from '@/services/cache'

export interface ThemeData {
  name: string
  font_family: string
  font_size: number
  bg_color: string
  fg_color: string
  highlight_color: string
  bg_image: string
}

export const themeStore = {
  defaultProperties: {
    '--color-fg': undefined,
    '--color-bg': undefined,
    '--color-highlight': undefined,
    '--bg-image': undefined,
    '--bg-position': undefined,
    '--bg-attachment': undefined,
    '--bg-size': undefined,
    '--font-family': undefined,
    '--font-size': undefined,
  } as Record<ThemeableProperty, string | undefined>,

  state: reactive({
    themes,
  }),

  init (theme: Theme | Theme['id'] = 'classic') {
    for (const key in this.defaultProperties) {
      this.defaultProperties[key] = document.body.style.getPropertyValue(key)
    }

    // calculate and set the highlight foreground color
    const observer = new StyleObserver(([{ value }]) => {
      document.body.style.setProperty(
        '--color-highlight-fg',
        Color(value).isDark() ? '#ffffff' : '#111111',
      )
    })

    observer.observe(document.body, '--color-highlight')

    if (typeof theme === 'object' && theme.is_custom) {
      // custom theme from server. Add it to the list of themes.
      this.state.themes.push(theme)
    }

    this.setTheme(theme)
  },

  get all () {
    return this.state.themes
  },

  setTheme (theme?: Theme | Theme['id']) {
    if (theme === undefined) {
      return this.setTheme(this.getCurrentTheme())
    }

    if (typeof theme === 'string') {
      theme = this.getThemeById(theme) ?? this.getDefaultTheme()
    }

    document.documentElement.setAttribute('data-theme', theme.id)
    const properties = Object.assign(clone(this.defaultProperties), theme.properties ?? {})

    for (const key in properties) {
      document.body.style.setProperty(key, properties[key]) // overriding :root
    }

    preferences.theme = theme.id
  },

  isCurrentTheme (theme: Theme | Theme['id']) {
    const currentTheme = this.getCurrentTheme()
    return typeof theme === 'string' ? currentTheme.id === theme : currentTheme.id === theme.id
  },

  getThemeById (id: Theme['id']) {
    return this.state.themes.find(theme => theme.id === id)
  },

  getDefaultTheme () {
    return this.getThemeById('classic')!
  },

  getCurrentTheme () {
    return preferences.theme
      ? (this.getThemeById(preferences.theme) ?? this.getDefaultTheme())
      : this.getDefaultTheme()
  },

  isValidTheme (id: Theme['id']) {
    return this.getThemeById(id) !== undefined
  },

  async store (data: ThemeData) {
    const theme = await http.post<Theme>('themes', data)
    this.state.themes.unshift(theme)
    cache.remove('custom-themes')

    return theme
  },

  async fetchCustomThemes () {
    const customThemes = await cache.remember('custom-themes', async () => await http.get<Theme[]>('themes'))
    this.state.themes = uniqBy(this.state.themes.concat(customThemes), 'id')
  },

  async destroy (theme: Theme) {
    if (!theme.is_custom) {
      return
    }

    const isCurrentTheme = this.isCurrentTheme(theme)

    await http.delete(`themes/${theme.id}`)
    this.state.themes = this.state.themes.filter(({ id }) => id !== theme.id)
    cache.remove('custom-themes')

    if (isCurrentTheme) {
      this.setTheme(this.getDefaultTheme())
    }
  },
}
