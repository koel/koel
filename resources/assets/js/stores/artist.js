import _ from 'lodash';

import config from '../config';
import albumStore from './album';
import sharedStore from './shared';

export default {
    state: {
        artists: [],
    },

    /**
     * Init the store.
     * 
     * @param  array artists The array of artists we got from the server.
     */
    init(artists = null) {
        this.state.artists = artists ? artists: sharedStore.state.artists;

        // Init the album store. This must be called prior to the next logic,
        // because we're using some data from the album store later.
        albumStore.init(this.state.artists);

        // Traverse through artists array to get the cover and number of songs for each.
        _.each(this.state.artists, artist => {
            this.getCover(artist);
            
            artist.songCount = _.reduce(artist.albums, (count, album)  => count + album.songs.length, 0);
        });
    },

    all() {
        return this.state.artists;
    },

    /**
     * Get all songs performed by an artist.
     *
     * @param object artist
     *
     * @return array
     */
    getSongsByArtist(artist) {
        if (!artist.songs) {
            artist.songs = _.reduce(artist.albums, (songs, album) => songs.concat(album.songs), []);
        }

        return artist.songs;
    },

    /**
     * Get the artist's cover
     *
     * @param object artist
     *
     * @return string
     */
    getCover(artist) {
        artist.cover = config.unknownCover;

        artist.albums.every(album => {
            // If there's a "real" cover, use it.
            if (album.cover != config.unknownCover) {
                artist.cover = album.cover;
                
                // I want to break free.
                return false;
            }
        });

        return artist.cover;
    },
};
