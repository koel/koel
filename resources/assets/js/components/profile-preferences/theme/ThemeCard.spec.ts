import { describe, expect, it } from 'vitest'
import { screen } from '@testing-library/vue'
import { createHarness } from '@/__tests__/TestHarness'
import Component from './ThemeCard.vue'
import { themeStore } from '@/stores/themeStore'

describe('themeCard.vue', () => {
  const h = createHarness()

  const renderComponent = () => {
    const theme = h.factory('theme', {
      name: 'Sample',
    })

    const rendered = h.render(Component, {
      props: {
        theme,
      },
    })

    return {
      ...rendered,
      theme,
    }
  }

  it('renders', () => expect(renderComponent().html()).toMatchSnapshot())

  it('sets the theme when clicked', async () => {
    const { theme } = renderComponent()
    const setThemeMock = h.mock(themeStore, 'setTheme')

    await h.user.click(screen.getByRole('button', { name: 'Sample' }))

    expect(setThemeMock).toHaveBeenCalledWith(theme)
  })
})
