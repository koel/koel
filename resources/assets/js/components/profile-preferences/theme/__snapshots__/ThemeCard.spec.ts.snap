// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`themeCard.vue > renders 1`] = `
<article data-v-a32b02e1="" title="Set current theme to Sample" class="theme h-[96px] bg-center bg-cover relative outline outline-1 outline-k-fg-10 cursor-pointer rounded-lg transition duration-300 shadow-lg" data-testid="theme-card">
  <!--v-if--><button data-v-a32b02e1="" class="opacity-0 hover:opacity-100 absolute h-full rounded-lg w-full top-0 left-0 flex items-center justify-center text-lg transition-opacity bg-black/20 text-white" type="button">Sample</button>
</article>
`;
